### ISSUE TITLE
Format only modified lines

### ISSUE THREAD

**Issue by kylegmaxwell** [Microsoft/vscode#44075]
### Issue Type
Feature Request

### Description
It would be helpful if there were a way to format only the modified lines. What is modified would be determined by source control. Currently formatOnType only works with semi colon (see https://github.com/Microsoft/vscode-cpptools/issues/1419), and formatOnSave formats the entire file. This is undesirable when working on large codebases with legacy code. In that case I want to make a small change, and have it automatically formatted on save, but not modify the rest of the file, which would create a huge diff.
It would also be great to have similar behavior for trimming trailing white space.

### VS Code Info
VS Code version: Code 1.20.1 (f88bbf9137d24d36d968ea6b2911786bfe103002, 2018-02-13T15:31:21.019Z)
OS version: Linux x64 4.14.16-200.fc26.x86_64

### Steps to Reproduce:
1. Set preferences:
    "editor.formatOnSave": true,
    "C_Cpp.clang_format_path": "/usr/bin/clang-format",
2. Open a C++ file with multiple formatting errors
3. Modify the file on one line.
4. Note that multiple modifications have been made to the file.

### Does this issue occur when all extensions are disabled?:
No, formatting does not work at all in that case.

**Comment by Gruntfuggly** [comment]
Been hoping for this for a while. In the meantime, I've knocked up an extension...

https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.format-modified

**Comment by kylegmaxwell** [comment]
Awesome! Just tried it out and it seems to work so far.

**Comment by alekseyt** [comment]
Great job, but this shouldn't be an extension IMO, this is a legit issue with editor and i don't personally think "feature-request" label is applicable here. I'm in the same boat and i have to use external merge tool (e.g.meld) to rollback unwanted changes.

**Comment by guillochon** [comment]
That extension appears to be C++ only, anyone know of a Python equivalent?

**Comment by devinrhode2** [comment]
Extension currently appears to support many languages now

**Comment by ronakg** [comment]
Any update on this issue? I don't want to use the extension mentioned above because all my formatting is set up via their respective language servers. This is a very valid use-case that needs addressing.

**Comment by jrieken** [comment]
Yes - if a project is not under source control and format on save is limited to modifications then no formatting happens. Tho, when a project is under source control than all modified and added lines/files are formatted.

**Comment by ronakg** [comment]
> Yes - if a project is not under source control and format on save is limited to modifications then no formatting happens. Tho, when a project is under source control than all modified and added lines/files are formatted.

I'm asking about a different setting `editor.formatOnType`, which is separate from `editor.formatOnSave`. Can you clarify how these 2 will work together now? I'm guessing `editor.formatOnType` will work as usual irrespective of `editor.formatOnSave`.

**Comment by jrieken** [comment]
Oh, misread that. Yeah, format on type has nothing to do with format on save and nothing has changed for that

**Comment by icharge** [comment]
How about for Format document by myself on modified code. Not just on save.

**Comment by jrieken** [comment]
It will only show up when you have a formatter that supports formatting modified ranges, e.g one that support range formatting. The rule of thumb is that if you have "Format Selection" then you also have "Format Modified Lines"

**Comment by MikeYuanMY** [comment]
which popular formatters currently support this? I have prettier but cannot seem working with this feature

**Comment by jrieken** [comment]
https://github.com/prettier/prettier-vscode/blob/main/README.md#format-selection

**Comment by icharge** [comment]
Oh I see, I'm using Prettier then I have to set new shortcut key. :)
<img width="721" alt="Screen Shot 2563-09-14 at 13 47 19" src="https://user-images.githubusercontent.com/1619007/93052694-d2c3eb00-f690-11ea-9bae-7bb85ca3811d.png">

**Comment by kylegmaxwell** [comment]
A lot of people are asking for format modified on type instead of save. I didn't ask for that since I thought it does not make sense. If you format on type then you always just format the lines as you modify them. I needed format modified since I wanted to make changes to several lines and then save and see just those lines formatted. Also I was using C++ with clang_format which does not work well with formatting single lines since it sometimes removes the line you are typing on, hence the desire to format modified lines on save.

**Comment by sheerun** [comment]
Unfortunately this often doesn't work often giving "Overlapping ranges are not allowed! error https://github.com/prettier/prettier-vscode/issues/497

**Comment by valeryan** [comment]
[My extension](https://github.com/valeryan/vscode-phpsab) uses PHPCS/CBF to format PHP files and because many of the rules are dealing with whitespace formatting allowing `format selection` does not work well. PHPCBF has an option to only format modified lines. I would like to be able to implement this without also implementing format by selection and just format the document using PHPCBF's internal filter for git modified. However, it seems that as far as I can tell the` format only modified lines` just piggybacks off of the already established API for format selection which I do not want to support. 

Also, because I do not implement `registerDocumentRangeFormattingEditProvider` my extension is not activated on save when a user now sets `Format on Save Mode` to modified. This is causing confusion because no errors or information is shown to the user to help them understand why my extension is not functioning in this mode.

**Comment by trentbullard** [comment]
not sure if this is the right place for this, but thank you for this @kylegmaxwell it was a sad day when i had to disable formatOnSave
